### PR TITLE
(feat): Validate Form Concepts and concept mappings

### PR DESCRIPTION
**Pull request by hadijahkyampeire**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary
- This adds form validation by getting all question-and-answer concepts and sending them to the backend and then checking if the concepts exist or are correct.
- Once the validation response is returned, if there are errors, they are then displayed below the question.
- Validation is hidden behind a `config` which is false by default, if turned true, a `validate` button will show next to the save form button.
- A form can be saved with validation errors but publishing can't be done unless validation passes.

## Screenshots

https://github.com/openmrs/openmrs-esm-form-builder/assets/30952856/f59d73d0-49b4-4d6f-abaa-43c7af10a31e

## Related Issue
- https://openmrs.atlassian.net/browse/O3-3149

### CODE REVIEW

File: src/components/action-buttons/action-buttons.component.tsx [hunk 1]
```diff
@@ -53,6 +75,19 @@ function ActionButtons({ schema, t }: ActionButtonsProps) {
...
   }
 
+  async function handleValidateAndPublish() {
+    setStatus('validateBeforePublishing');
+    const [errorsArray] = await handleFormValidation(schema, dataTypeToRenderingMap);
+    setValidationResponse(errorsArray);
+    if (errorsArray.length > 0) {
+      setStatus('validated');
+      setValidationComplete(true);
+      setPublishedWithErrors(true);
```
Comment (arodidev) [inline src/components/action-buttons/action-buttons.component.tsx:85]: Does this mean we are going to let the user publish the form if it has validation errors?

File: src/components/action-buttons/action-buttons.component.tsx [hunk 2]
```diff
@@ -53,6 +75,19 @@ function ActionButtons({ schema, t }: ActionButtonsProps) {
     }
   }
 
+  async function handleValidateAndPublish() {
+    setStatus('validateBeforePublishing');
+    const [errorsArray] = await handleFormValidation(schema, dataTypeToRenderingMap);
+    setValidationResponse(errorsArray);
+    if (errorsArray.length > 0) {
```
Comment (arodidev) [inline src/components/action-buttons/action-buttons.component.tsx:82]:
```suggestion
    if (errorsArray.length) {
```

File: src/components/action-buttons/action-buttons.component.tsx (hunk 1 above)
Comment (hadijahkyampeire) [reply src/components/action-buttons/action-buttons.component.tsx:85]: No actually, it means that set a state that the user is trying to publish with errors, I need that state to show some inlineNotification.

File: src/components/interactive-builder/interactive-builder.component.tsx [hunk 3]
```diff
@@ -269,6 +281,19 @@ const InteractiveBuilder: React.FC<InteractiveBuilderProps> = ({ isLoading, onSc
     setShowDeleteQuestionModal(true);
   };
 
+  const getAnswerErrors = (answers: Array<Record<string, string>>) => {
+    const answerLabels = answers?.map((a) => a.label) || [];
+    const errors: Array<ValidationError> = validationResponse.filter((e) => answerLabels?.includes(e.field.label));
```
Comment (arodidev) [inline src/components/interactive-builder/interactive-builder.component.tsx:286]:
```suggestion
    const errors: Array<ValidationError> = validationResponse.filter((error) => answerLabels?.includes(error.field.label));
```

File: src/components/interactive-builder/interactive-builder.component.tsx [hunk 4]
```diff
@@ -269,6 +281,19 @@ const InteractiveBuilder: React.FC<InteractiveBuilderProps> = ({ isLoading, onSc
     setShowDeleteQuestionModal(true);
   };
 
+  const getAnswerErrors = (answers: Array<Record<string, string>>) => {
+    const answerLabels = answers?.map((a) => a.label) || [];
```
Comment (arodidev) [inline src/components/interactive-builder/interactive-builder.component.tsx:285]:
```suggestion
    const answerLabels = answers?.map((answer) => answer.label) || [];
```

File: src/components/interactive-builder/interactive-builder.component.tsx (hunk 3 above)
Comment (arodidev) [reply src/components/interactive-builder/interactive-builder.component.tsx:286]: For ease of readability.

File: src/components/interactive-builder/interactive-builder.component.tsx [hunk 5]
```diff
@@ -269,6 +281,19 @@ const InteractiveBuilder: React.FC<InteractiveBuilderProps> = ({ isLoading, onSc
...
+    const errors: Array<ValidationError> = validationResponse.filter((e) => answerLabels?.includes(e.field.label));
+    return errors || [];
+  };
+
+  const getValidationError = (question: Question) => {
+    const errorField: ValidationError = validationResponse.find(
+      (e) => e.field.label === question.label && e.field.id === question.id && e.field.type === question.type,
```
Comment (arodidev) [inline src/components/interactive-builder/interactive-builder.component.tsx:292]: Here as well.

File: src/components/interactive-builder/interactive-builder.component.tsx [hunk 6]
```diff
@@ -269,6 +281,19 @@ const InteractiveBuilder: React.FC<InteractiveBuilderProps> = ({ isLoading, onSc
...
+    );
+    return errorField?.errorMessage || errorField?.warningMessage || '';
+  };
```
Comment (arodidev) [inline src/components/interactive-builder/interactive-builder.component.tsx:295]: I think for now we can just do the error message and leave out the warning message, these might clog the UI and we may not need to represent them if they are not deal-breakers for form-submission. What do you think?

File: src/config-schema.ts [hunk 7]
```diff
@@ -41,4 +41,22 @@ export const configSchema = {
...
+    _default: {
+      Numeric: ['number', 'fixed-value'],
+      Coded: ['select', 'checkbox', 'radio', 'toggle', 'content-switcher', 'fixed-value'],
+      Text: ['text', 'textarea', 'fixed-value'],
+      Date: ['date', 'fixed-value'],
+      Datetime: ['datetime', 'fixed-value'],
+      Boolean: ['toggle', 'select', 'radio', 'content-switcher', 'fixed-value'],
+      Rule: ['repeating', 'group'],
+    },
+  },
```
Comment (arodidev) [inline src/config-schema.ts:56]: There is a bit technical debt here and might be addressed in a separate PR, but we may need to update this map to include more rendering types that may be supported by the various concept datatypes.

**Review by arodidev — COMMENTED** [review]
Thanks for this, @hadijahkyampeire. LGTM largely, though I left a few minor comments.

File: src/components/interactive-builder/interactive-builder.component.tsx (hunk 6 above)
Comment (hadijahkyampeire) [reply src/components/interactive-builder/interactive-builder.component.tsx:295]: @arodidev I have actually not seen a form with `warningMessages` but just to ensure that if they are returned they don't break the UI, I am just returning them if they exist.

**Review by arodidev — APPROVED** [review]

**Comment by hadijahkyampeire** [comment]
Thanks @arodidev
